src/box/sql/mem.c

```diff
@@ -988,6 +1023,69 @@ double_to_uint_forced(struct Mem *mem)
 	return res;
 }
 
+static inline int
+double_to_dec(struct Mem *mem)
+{
+	assert(mem->type == MEM_TYPE_DOUBLE);
+	double d = mem->u.r;
+	if (d >= 1e38 || d <= -1e38)
+		return -1;
+	decimal_from_double(&mem->u.d, d);
+	mem->type = MEM_TYPE_DEC;
+	mem->flags = 0;
+	return 0;
+}
+
+static inline int
+double_to_dec_precise(struct Mem *mem)
+{
+	assert(mem->type == MEM_TYPE_DOUBLE);
+	double d = mem->u.r;
+	if (d >= 1e38 || d <= -1e38)
+		return -1;
+	decimal_t dec;
+	decimal_from_double(&dec, d);
+	if (atof(decimal_str(&mem->u.d)) != d)
+		return -1;
+	mem->u.d = dec;
+	mem->type = MEM_TYPE_DEC;
+	mem->flags = 0;
+	return 0;
+}
+
+/**
+ * Cast MEM with DOUBLE to DECIMAL. Doesn't fail. The return value is < 0 if
+ * the original value is less than the result, > 0 if the original value is
+ * greater than the result, and 0 if the cast is precise.
+ */
+static inline int
+double_to_dec_forced(struct Mem *mem)
+{
+	assert(mem->type == MEM_TYPE_DOUBLE);
+	double d = mem->u.r;
+	mem->type = MEM_TYPE_DEC;
+	mem->flags = 0;
+	if (d >= 1e38) {
+		const char *val = "99999999999999999999999999999999999999";
+		assert(strlen(val) == 38);
+		decimal_from_string(&mem->u.d, val);
```

Comment:
These conversions from string looks quite strange. Could we expand somehow decimal API and add direct method to achieve this?

Reply:
I think of it a bit later. @tsafin asked @sergepetrenko, who said this:

>I agree, it would be best to introduce decimal_to_double() to src/lib/core/decimal
>
>I can't think of an easy way here. The only one I see is construct the double value
>digit by digit. This mustn't be too different from decimal->string->double conversion. 

I plan to leave it as it is, for now. I still plan to fix this later.